lib/backend/memory/memory.go

```diff
@@ -290,25 +302,105 @@ func (m *Memory) DeleteRange(ctx context.Context, startKey, endKey backend.Key)
 	return nil
 }
 
-// GetRange returns query range
-func (m *Memory) GetRange(ctx context.Context, startKey, endKey backend.Key, limit int) (*backend.GetResult, error) {
-	if startKey.IsZero() {
-		return nil, trace.BadParameter("missing parameter startKey")
+func (m *Memory) Items(ctx context.Context, params backend.IterateParams) iter.Seq2[backend.Item, error] {
+	if params.StartKey.IsZero() {
+		err := trace.BadParameter("missing parameter startKey")
+		return func(yield func(backend.Item, error) bool) { yield(backend.Item{}, err) }
 	}
-	if endKey.IsZero() {
-		return nil, trace.BadParameter("missing parameter endKey")
+	if params.EndKey.IsZero() {
+		err := trace.BadParameter("missing parameter endKey")
+		return func(yield func(backend.Item, error) bool) { yield(backend.Item{}, err) }
 	}
+
+	limit := params.Limit
 	if limit <= 0 {
 		limit = backend.DefaultRangeLimit
 	}
-	m.Lock()
-	defer m.Unlock()
-	m.removeExpired()
-	re := m.getRange(ctx, startKey, endKey, limit)
-	if len(re.Items) == backend.DefaultRangeLimit {
+
+	const defaultPageSize = 1000
+	return func(yield func(backend.Item, error) bool) {
+		startKey := params.StartKey
+		endKey := params.EndKey
+		compareDirection := 1
+		itemIter := m.tree.AscendGreaterOrEqual
+		if params.Descending {
+			startKey = params.EndKey
+			endKey = params.StartKey
+			compareDirection = -1
+			itemIter = m.tree.DescendLessOrEqual
+		}
+
+		btreeItems := func(start *btreeItem) iter.Seq[*btreeItem] {
+			return func(yield func(*btreeItem) bool) {
+				m.Lock()
+				defer m.Unlock()
+				m.removeExpired()
+
+				itemIter(start, yield)
+			}
+		}
+
+		var excludedStart bool
+		var totalCount int
+		items := make([]backend.Item, 0, min(limit, defaultPageSize))
+		startItem := &btreeItem{Item: backend.Item{Key: startKey}}
+		for {
+			pageLimit := min(limit-totalCount, defaultPageSize)
+			items = items[:0]
+
+			for item := range btreeItems(startItem) {
+				if item.Key.Compare(endKey)*compareDirection > 0 {
+					startItem = item
+					break
+				}
+
+				if excludedStart {
+					excludedStart = false
+					if item.Key.Compare(startItem.Key) == 0 {
+						continue
+					}
+				}
+
+				items = append(items, item.Item)
+				if len(items) >= pageLimit {
+					startItem = item
```

Comment:
Do we need to update `startItem` here? Can't we take it from `items[len(items)-1]` later like we do in other backends? Same goes for `excludedStart` since it can just always be set if we're reentering the outer loop.